From d473cab6f4dd3dc27dd70917f84ede7db59c63aa Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Sun, 28 Mar 2021 17:57:27 -0500
Subject: [PATCH 059/196] Always require PrecomputedTransactionData in
 GenericTransactionSignatureChecker. (#109)

Basically deletes the constructor which sets txdata to nullptr.

This change is required for BIP341 sighashes, as PrecomputedTransactionData contains m_spent_outputs, which is required for computing parts of the sighash.
---
 src/bitcoin-tx.cpp             |   2 +-
 src/rpc/rawtransaction.cpp     |   3 +-
 src/script/interpreter.h       |   3 -
 src/script/sign.cpp            |  12 ++-
 src/script/sign.h              |   3 +-
 src/test/multisig_tests.cpp    |  63 +++++++-------
 src/test/script_p2sh_tests.cpp |  10 +--
 src/test/script_tests.cpp      | 149 +++++++++++++++++----------------
 src/test/transaction_tests.cpp |   9 +-
 src/wallet/wallet.cpp          |  19 ++++-
 10 files changed, 152 insertions(+), 121 deletions(-)

diff --git a/src/bitcoin-tx.cpp b/src/bitcoin-tx.cpp
index 87d09f6c1..ed10a029d 100644
--- a/src/bitcoin-tx.cpp
+++ b/src/bitcoin-tx.cpp
@@ -703,7 +703,7 @@ static void MutateTxSign(CMutableTransaction &tx, const std::string &flagStr) {
         const Amount amount = coin.GetTxOut().nValue;
 
         SignatureData sigdata =
-            DataFromTransaction(mergedTx, i, coin.GetTxOut());
+            DataFromTransaction(mergedTx, i, txdata);
         // Only sign SIGHASH_SINGLE if there's a corresponding output:
         if ((sigHashType.getBaseType() != BaseSigHashType::SINGLE) ||
             (i < mergedTx.vout.size())) {
diff --git a/src/rpc/rawtransaction.cpp b/src/rpc/rawtransaction.cpp
index c6dd32440..7e539f607 100644
--- a/src/rpc/rawtransaction.cpp
+++ b/src/rpc/rawtransaction.cpp
@@ -807,7 +807,8 @@ static UniValue combinerawtransaction(const Config &config,
         // ... and merge in other signatures:
         for (const CMutableTransaction &txv : txVariants) {
             if (txv.vin.size() > i) {
-                sigdata.MergeSignatureData(DataFromTransaction(txv, i, txout));
+                sigdata.MergeSignatureData(
+                    DataFromTransaction(txv, i, txdata));
             }
         }
         ProduceSignature(DUMMY_SIGNING_PROVIDER,
diff --git a/src/script/interpreter.h b/src/script/interpreter.h
index fdf6d6c7a..aa17010e1 100644
--- a/src/script/interpreter.h
+++ b/src/script/interpreter.h
@@ -61,9 +61,6 @@ private:
     const PrecomputedTransactionData *txdata;
 
 public:
-    GenericTransactionSignatureChecker(const T *txToIn, unsigned int nInIn,
-                                       const Amount &amountIn)
-        : txTo(txToIn), nIn(nInIn), amount(amountIn), txdata(nullptr) {}
     GenericTransactionSignatureChecker(
         const T *txToIn, unsigned int nInIn, const Amount &amountIn,
         const PrecomputedTransactionData &txdataIn)
diff --git a/src/script/sign.cpp b/src/script/sign.cpp
index 62e4c8dd3..a9c4e533b 100644
--- a/src/script/sign.cpp
+++ b/src/script/sign.cpp
@@ -275,14 +275,17 @@ struct Stacks {
 // Extracts signatures and scripts from incomplete scriptSigs. Please do not
 // extend this, use PSBT instead
 SignatureData DataFromTransaction(const CMutableTransaction &tx,
-                                  unsigned int nIn, const CTxOut &txout) {
+                                  unsigned int nIn,
+                                  const PrecomputedTransactionData &txdata) {
     SignatureData data;
     assert(tx.vin.size() > nIn);
     data.scriptSig = tx.vin[nIn].scriptSig;
     Stacks stack(data);
+    const CTxOut &txout = txdata.m_spent_outputs[nIn];
 
     // Get signatures
-    MutableTransactionSignatureChecker tx_checker(&tx, nIn, txout.nValue);
+    MutableTransactionSignatureChecker tx_checker(&tx, nIn, txout.nValue,
+                                                  txdata);
     SignatureExtractorChecker extractor_checker(data, tx_checker);
     if (VerifyScript(data.scriptSig, txout.scriptPubKey,
                      STANDARD_SCRIPT_VERIFY_FLAGS, extractor_checker)) {
@@ -484,7 +487,7 @@ bool SignTransaction(CMutableTransaction &mtx, const SigningProvider *keystore,
         const CScript &prevPubKey = coinTxOut.scriptPubKey;
         const Amount amount = coinTxOut.nValue;
 
-        SignatureData sigdata = DataFromTransaction(mtx, i, coinTxOut);
+        SignatureData sigdata = DataFromTransaction(mtx, i, txdata);
         // Only sign SIGHASH_SINGLE if there's a corresponding output:
         if ((sigHashType.getBaseType() != BaseSigHashType::SINGLE) ||
             (i < mtx.vout.size())) {
@@ -505,7 +508,8 @@ bool SignTransaction(CMutableTransaction &mtx, const SigningProvider *keystore,
         ScriptError serror = ScriptError::OK;
         if (!VerifyScript(
                 txin.scriptSig, prevPubKey, STANDARD_SCRIPT_VERIFY_FLAGS,
-                TransactionSignatureChecker(&txConst, i, amount), &serror)) {
+                TransactionSignatureChecker(&txConst, i, amount, txdata),
+                &serror)) {
             if (serror == ScriptError::INVALID_STACK_OPERATION) {
                 // Unable to sign input and verification failed (possible
                 // attempt to partially sign).
diff --git a/src/script/sign.h b/src/script/sign.h
index ed6cc5002..830729208 100644
--- a/src/script/sign.h
+++ b/src/script/sign.h
@@ -184,7 +184,8 @@ bool SignSignature(const SigningProvider &provider, const CTransaction &txFrom,
 
 /** Extract signature data from a transaction input, and insert it. */
 SignatureData DataFromTransaction(const CMutableTransaction &tx,
-                                  unsigned int nIn, const CTxOut &txout);
+                                  unsigned int nIn,
+                                  const PrecomputedTransactionData &txdata);
 void UpdateInput(CTxIn &input, const SignatureData &data);
 
 /**
diff --git a/src/test/multisig_tests.cpp b/src/test/multisig_tests.cpp
index a9bd0027f..dee2290ad 100644
--- a/src/test/multisig_tests.cpp
+++ b/src/test/multisig_tests.cpp
@@ -70,11 +70,13 @@ BOOST_AUTO_TEST_CASE(multisig_verify) {
 
     // Spending transaction
     CMutableTransaction txTo[3];
+    std::vector<PrecomputedTransactionData> txdata;
     for (int i = 0; i < 3; i++) {
         txTo[i].vin.resize(1);
         txTo[i].vout.resize(1);
         txTo[i].vin[0].prevout = COutPoint(txFrom.GetId(), i);
         txTo[i].vout[0].nValue = SATOSHI;
+        txdata.push_back(PrecomputedTransactionData(txTo[i], {txFrom.vout[i]}));
     }
 
     std::vector<CKey> keys;
@@ -86,28 +88,29 @@ BOOST_AUTO_TEST_CASE(multisig_verify) {
     s = sign_multisig(a_and_b, keys, txTo[0], 0);
     BOOST_CHECK(VerifyScript(
         s, a_and_b, flags,
-        MutableTransactionSignatureChecker(&txTo[0], 0, amount), &err));
+        MutableTransactionSignatureChecker(&txTo[0], 0, amount, txdata[0]),
+        &err));
     BOOST_CHECK_MESSAGE(err == ScriptError::OK, ScriptErrorString(err));
 
     for (int i = 0; i < 4; i++) {
         keys.assign(1, key[i]);
         s = sign_multisig(a_and_b, keys, txTo[0], 0);
-        BOOST_CHECK_MESSAGE(
-            !VerifyScript(
-                s, a_and_b, flags,
-                MutableTransactionSignatureChecker(&txTo[0], 0, amount), &err),
-            strprintf("a&b 1: %d", i));
+        BOOST_CHECK_MESSAGE(!VerifyScript(s, a_and_b, flags,
+                                          MutableTransactionSignatureChecker(
+                                              &txTo[0], 0, amount, txdata[0]),
+                                          &err),
+                            strprintf("a&b 1: %d", i));
         BOOST_CHECK_MESSAGE(err == ScriptError::INVALID_STACK_OPERATION,
                             ScriptErrorString(err));
 
         keys.assign(1, key[1]);
         keys.push_back(key[i]);
         s = sign_multisig(a_and_b, keys, txTo[0], 0);
-        BOOST_CHECK_MESSAGE(
-            !VerifyScript(
-                s, a_and_b, flags,
-                MutableTransactionSignatureChecker(&txTo[0], 0, amount), &err),
-            strprintf("a&b 2: %d", i));
+        BOOST_CHECK_MESSAGE(!VerifyScript(s, a_and_b, flags,
+                                          MutableTransactionSignatureChecker(
+                                              &txTo[0], 0, amount, txdata[0]),
+                                          &err),
+                            strprintf("a&b 2: %d", i));
         BOOST_CHECK_MESSAGE(err == ScriptError::SIG_NULLFAIL,
                             ScriptErrorString(err));
     }
@@ -117,18 +120,19 @@ BOOST_AUTO_TEST_CASE(multisig_verify) {
         keys.assign(1, key[i]);
         s = sign_multisig(a_or_b, keys, txTo[1], 0);
         if (i == 0 || i == 1) {
-            BOOST_CHECK_MESSAGE(VerifyScript(s, a_or_b, flags,
-                                             MutableTransactionSignatureChecker(
-                                                 &txTo[1], 0, amount),
-                                             &err),
-                                strprintf("a|b: %d", i));
+            BOOST_CHECK_MESSAGE(
+                VerifyScript(s, a_or_b, flags,
+                             MutableTransactionSignatureChecker(
+                                 &txTo[1], 0, amount, txdata[1]),
+                             &err),
+                strprintf("a|b: %d", i));
             BOOST_CHECK_MESSAGE(err == ScriptError::OK, ScriptErrorString(err));
         } else {
             BOOST_CHECK_MESSAGE(
-                !VerifyScript(
-                    s, a_or_b, flags,
-                    MutableTransactionSignatureChecker(&txTo[1], 0, amount),
-                    &err),
+                !VerifyScript(s, a_or_b, flags,
+                              MutableTransactionSignatureChecker(
+                                  &txTo[1], 0, amount, txdata[1]),
+                              &err),
                 strprintf("a|b: %d", i));
             BOOST_CHECK_MESSAGE(err == ScriptError::SIG_NULLFAIL,
                                 ScriptErrorString(err));
@@ -138,7 +142,8 @@ BOOST_AUTO_TEST_CASE(multisig_verify) {
     s << OP_0 << OP_1;
     BOOST_CHECK(!VerifyScript(
         s, a_or_b, flags,
-        MutableTransactionSignatureChecker(&txTo[1], 0, amount), &err));
+        MutableTransactionSignatureChecker(&txTo[1], 0, amount, txdata[1]),
+        &err));
     BOOST_CHECK_MESSAGE(err == ScriptError::SIG_DER, ScriptErrorString(err));
 
     for (int i = 0; i < 4; i++) {
@@ -148,19 +153,19 @@ BOOST_AUTO_TEST_CASE(multisig_verify) {
             s = sign_multisig(escrow, keys, txTo[2], 0);
             if (i < j && i < 3 && j < 3) {
                 BOOST_CHECK_MESSAGE(
-                    VerifyScript(
-                        s, escrow, flags,
-                        MutableTransactionSignatureChecker(&txTo[2], 0, amount),
-                        &err),
+                    VerifyScript(s, escrow, flags,
+                                 MutableTransactionSignatureChecker(
+                                     &txTo[2], 0, amount, txdata[2]),
+                                 &err),
                     strprintf("escrow 1: %d %d", i, j));
                 BOOST_CHECK_MESSAGE(err == ScriptError::OK,
                                     ScriptErrorString(err));
             } else {
                 BOOST_CHECK_MESSAGE(
-                    !VerifyScript(
-                        s, escrow, flags,
-                        MutableTransactionSignatureChecker(&txTo[2], 0, amount),
-                        &err),
+                    !VerifyScript(s, escrow, flags,
+                                  MutableTransactionSignatureChecker(
+                                      &txTo[2], 0, amount, txdata[2]),
+                                  &err),
                     strprintf("escrow 2: %d %d", i, j));
                 BOOST_CHECK_MESSAGE(err == ScriptError::SIG_NULLFAIL,
                                     ScriptErrorString(err));
diff --git a/src/test/script_p2sh_tests.cpp b/src/test/script_p2sh_tests.cpp
index f08bd1e56..5fc3f8876 100644
--- a/src/test/script_p2sh_tests.cpp
+++ b/src/test/script_p2sh_tests.cpp
@@ -39,11 +39,11 @@ static bool Verify(const CScript &scriptSig, const CScript &scriptPubKey,
     txTo.vin[0].scriptSig = scriptSig;
     txTo.vout[0].nValue = SATOSHI;
 
-    return VerifyScript(
-        scriptSig, scriptPubKey,
-        SCRIPT_ENABLE_SIGHASH_FORKID,
-        MutableTransactionSignatureChecker(&txTo, 0, txFrom.vout[0].nValue),
-        &err);
+    PrecomputedTransactionData txdata(txTo, std::vector(txTo.vout));
+    return VerifyScript(scriptSig, scriptPubKey, SCRIPT_ENABLE_SIGHASH_FORKID,
+                        MutableTransactionSignatureChecker(
+                            &txTo, 0, txFrom.vout[0].nValue, txdata),
+                        &err);
 }
 
 BOOST_FIXTURE_TEST_SUITE(script_p2sh_tests, BasicTestingSetup)
diff --git a/src/test/script_tests.cpp b/src/test/script_tests.cpp
index 11f0a522d..ecccf21ae 100644
--- a/src/test/script_tests.cpp
+++ b/src/test/script_tests.cpp
@@ -131,11 +131,13 @@ static void DoTest(const CScript &scriptPubKey, const CScript &scriptSig,
         BuildCreditingTransaction(scriptPubKey, nValue)};
     CMutableTransaction tx = BuildSpendingTransaction(scriptSig, txCredit);
     CMutableTransaction tx2 = tx;
-    BOOST_CHECK_MESSAGE(VerifyScript(scriptSig, scriptPubKey, flags,
-                                     MutableTransactionSignatureChecker(
-                                         &tx, 0, txCredit.vout[0].nValue),
-                                     &err) == expect,
-                        message);
+    const PrecomputedTransactionData txdata(tx, std::vector(txCredit.vout));
+    BOOST_CHECK_MESSAGE(
+        VerifyScript(scriptSig, scriptPubKey, flags,
+                     MutableTransactionSignatureChecker(
+                         &tx, 0, txCredit.vout[0].nValue, txdata),
+                     &err) == expect,
+        message);
     BOOST_CHECK_MESSAGE(err == scriptError, FormatScriptError(err) + " where " +
                                                 FormatScriptError(scriptError) +
                                                 " expected: " + message);
@@ -151,14 +153,14 @@ static void DoTest(const CScript &scriptPubKey, const CScript &scriptSig,
         uint32_t combined_flags =
             expect ? (flags & ~extra_flags) : (flags | extra_flags);
 
-        BOOST_CHECK_MESSAGE(VerifyScript(scriptSig, scriptPubKey,
-                                         combined_flags,
-                                         MutableTransactionSignatureChecker(
-                                             &tx, 0, txCredit.vout[0].nValue),
-                                         &err) == expect,
-                            message + strprintf(" (with %s flags %08x)",
-                                                expect ? "removed" : "added",
-                                                combined_flags ^ flags));
+        BOOST_CHECK_MESSAGE(
+            VerifyScript(scriptSig, scriptPubKey, combined_flags,
+                         MutableTransactionSignatureChecker(
+                             &tx, 0, txCredit.vout[0].nValue, txdata),
+                         &err) == expect,
+            message + strprintf(" (with %s flags %08x)",
+                                expect ? "removed" : "added",
+                                combined_flags ^ flags));
     }
 
 #if defined(HAVE_CONSENSUS_LIB)
@@ -2198,35 +2200,36 @@ BOOST_AUTO_TEST_CASE(script_CHECKMULTISIG12) {
     const CTransaction txFrom12{
         BuildCreditingTransaction(scriptPubKey12, Amount::zero())};
     CMutableTransaction txTo12 = BuildSpendingTransaction(CScript(), txFrom12);
+    PrecomputedTransactionData txdata(txTo12, std::vector(txFrom12.vout));
 
     CScript goodsig1 =
         sign_multisig(scriptPubKey12, key1, CTransaction(txTo12));
-    BOOST_CHECK(VerifyScript(
-        goodsig1, scriptPubKey12, gFlags,
-        MutableTransactionSignatureChecker(&txTo12, 0, txFrom12.vout[0].nValue),
-        &err));
+    BOOST_CHECK(VerifyScript(goodsig1, scriptPubKey12, gFlags,
+                             MutableTransactionSignatureChecker(
+                                 &txTo12, 0, txFrom12.vout[0].nValue, txdata),
+                             &err));
     BOOST_CHECK_MESSAGE(err == ScriptError::OK, ScriptErrorString(err));
     txTo12.vout[0].nValue = 2 * SATOSHI;
-    BOOST_CHECK(!VerifyScript(
-        goodsig1, scriptPubKey12, gFlags,
-        MutableTransactionSignatureChecker(&txTo12, 0, txFrom12.vout[0].nValue),
-        &err));
+    BOOST_CHECK(!VerifyScript(goodsig1, scriptPubKey12, gFlags,
+                              MutableTransactionSignatureChecker(
+                                  &txTo12, 0, txFrom12.vout[0].nValue, txdata),
+                              &err));
     BOOST_CHECK_MESSAGE(err == ScriptError::SIG_NULLFAIL,
                         ScriptErrorString(err));
 
     CScript goodsig2 =
         sign_multisig(scriptPubKey12, key2, CTransaction(txTo12));
-    BOOST_CHECK(VerifyScript(
-        goodsig2, scriptPubKey12, gFlags,
-        MutableTransactionSignatureChecker(&txTo12, 0, txFrom12.vout[0].nValue),
-        &err));
+    BOOST_CHECK(VerifyScript(goodsig2, scriptPubKey12, gFlags,
+                             MutableTransactionSignatureChecker(
+                                 &txTo12, 0, txFrom12.vout[0].nValue, txdata),
+                             &err));
     BOOST_CHECK_MESSAGE(err == ScriptError::OK, ScriptErrorString(err));
 
     CScript badsig1 = sign_multisig(scriptPubKey12, key3, CTransaction(txTo12));
-    BOOST_CHECK(!VerifyScript(
-        badsig1, scriptPubKey12, gFlags,
-        MutableTransactionSignatureChecker(&txTo12, 0, txFrom12.vout[0].nValue),
-        &err));
+    BOOST_CHECK(!VerifyScript(badsig1, scriptPubKey12, gFlags,
+                              MutableTransactionSignatureChecker(
+                                  &txTo12, 0, txFrom12.vout[0].nValue, txdata),
+                              &err));
     BOOST_CHECK_MESSAGE(err == ScriptError::SIG_NULLFAIL,
                         ScriptErrorString(err));
 }
@@ -2254,45 +2257,46 @@ BOOST_AUTO_TEST_CASE(script_CHECKMULTISIG23) {
     // we can convert it to readonly transaction and use
     // TransactionSignatureChecker instead of MutableTransactionSignatureChecker
     const CTransaction txTo23(mutableTxTo23);
+    const PrecomputedTransactionData txdata(txTo23, std::vector(txFrom23.vout));
 
     std::vector<CKey> keys;
     keys.push_back(key1);
     keys.push_back(key2);
     CScript goodsig1 = sign_multisig(scriptPubKey23, keys, txTo23);
-    BOOST_CHECK(VerifyScript(
-        goodsig1, scriptPubKey23, gFlags,
-        TransactionSignatureChecker(&txTo23, 0, txFrom23.vout[0].nValue),
-        &err));
+    BOOST_CHECK(VerifyScript(goodsig1, scriptPubKey23, gFlags,
+                             TransactionSignatureChecker(
+                                 &txTo23, 0, txFrom23.vout[0].nValue, txdata),
+                             &err));
     BOOST_CHECK_MESSAGE(err == ScriptError::OK, ScriptErrorString(err));
 
     keys.clear();
     keys.push_back(key1);
     keys.push_back(key3);
     CScript goodsig2 = sign_multisig(scriptPubKey23, keys, txTo23);
-    BOOST_CHECK(VerifyScript(
-        goodsig2, scriptPubKey23, gFlags,
-        TransactionSignatureChecker(&txTo23, 0, txFrom23.vout[0].nValue),
-        &err));
+    BOOST_CHECK(VerifyScript(goodsig2, scriptPubKey23, gFlags,
+                             TransactionSignatureChecker(
+                                 &txTo23, 0, txFrom23.vout[0].nValue, txdata),
+                             &err));
     BOOST_CHECK_MESSAGE(err == ScriptError::OK, ScriptErrorString(err));
 
     keys.clear();
     keys.push_back(key2);
     keys.push_back(key3);
     CScript goodsig3 = sign_multisig(scriptPubKey23, keys, txTo23);
-    BOOST_CHECK(VerifyScript(
-        goodsig3, scriptPubKey23, gFlags,
-        TransactionSignatureChecker(&txTo23, 0, txFrom23.vout[0].nValue),
-        &err));
+    BOOST_CHECK(VerifyScript(goodsig3, scriptPubKey23, gFlags,
+                             TransactionSignatureChecker(
+                                 &txTo23, 0, txFrom23.vout[0].nValue, txdata),
+                             &err));
     BOOST_CHECK_MESSAGE(err == ScriptError::OK, ScriptErrorString(err));
 
     keys.clear();
     keys.push_back(key2);
     keys.push_back(key2); // Can't re-use sig
     CScript badsig1 = sign_multisig(scriptPubKey23, keys, txTo23);
-    BOOST_CHECK(!VerifyScript(
-        badsig1, scriptPubKey23, gFlags,
-        TransactionSignatureChecker(&txTo23, 0, txFrom23.vout[0].nValue),
-        &err));
+    BOOST_CHECK(!VerifyScript(badsig1, scriptPubKey23, gFlags,
+                              TransactionSignatureChecker(
+                                  &txTo23, 0, txFrom23.vout[0].nValue, txdata),
+                              &err));
     BOOST_CHECK_MESSAGE(err == ScriptError::SIG_NULLFAIL,
                         ScriptErrorString(err));
 
@@ -2300,10 +2304,10 @@ BOOST_AUTO_TEST_CASE(script_CHECKMULTISIG23) {
     keys.push_back(key2);
     keys.push_back(key1); // sigs must be in correct order
     CScript badsig2 = sign_multisig(scriptPubKey23, keys, txTo23);
-    BOOST_CHECK(!VerifyScript(
-        badsig2, scriptPubKey23, gFlags,
-        TransactionSignatureChecker(&txTo23, 0, txFrom23.vout[0].nValue),
-        &err));
+    BOOST_CHECK(!VerifyScript(badsig2, scriptPubKey23, gFlags,
+                              TransactionSignatureChecker(
+                                  &txTo23, 0, txFrom23.vout[0].nValue, txdata),
+                              &err));
     BOOST_CHECK_MESSAGE(err == ScriptError::SIG_NULLFAIL,
                         ScriptErrorString(err));
 
@@ -2311,10 +2315,10 @@ BOOST_AUTO_TEST_CASE(script_CHECKMULTISIG23) {
     keys.push_back(key3);
     keys.push_back(key2); // sigs must be in correct order
     CScript badsig3 = sign_multisig(scriptPubKey23, keys, txTo23);
-    BOOST_CHECK(!VerifyScript(
-        badsig3, scriptPubKey23, gFlags,
-        TransactionSignatureChecker(&txTo23, 0, txFrom23.vout[0].nValue),
-        &err));
+    BOOST_CHECK(!VerifyScript(badsig3, scriptPubKey23, gFlags,
+                              TransactionSignatureChecker(
+                                  &txTo23, 0, txFrom23.vout[0].nValue, txdata),
+                              &err));
     BOOST_CHECK_MESSAGE(err == ScriptError::SIG_NULLFAIL,
                         ScriptErrorString(err));
 
@@ -2322,10 +2326,10 @@ BOOST_AUTO_TEST_CASE(script_CHECKMULTISIG23) {
     keys.push_back(key4);
     keys.push_back(key2); // sigs must match pubkeys
     CScript badsig4 = sign_multisig(scriptPubKey23, keys, txTo23);
-    BOOST_CHECK(!VerifyScript(
-        badsig4, scriptPubKey23, gFlags,
-        TransactionSignatureChecker(&txTo23, 0, txFrom23.vout[0].nValue),
-        &err));
+    BOOST_CHECK(!VerifyScript(badsig4, scriptPubKey23, gFlags,
+                              TransactionSignatureChecker(
+                                  &txTo23, 0, txFrom23.vout[0].nValue, txdata),
+                              &err));
     BOOST_CHECK_MESSAGE(err == ScriptError::SIG_NULLFAIL,
                         ScriptErrorString(err));
 
@@ -2333,19 +2337,19 @@ BOOST_AUTO_TEST_CASE(script_CHECKMULTISIG23) {
     keys.push_back(key1);
     keys.push_back(key4); // sigs must match pubkeys
     CScript badsig5 = sign_multisig(scriptPubKey23, keys, txTo23);
-    BOOST_CHECK(!VerifyScript(
-        badsig5, scriptPubKey23, gFlags,
-        TransactionSignatureChecker(&txTo23, 0, txFrom23.vout[0].nValue),
-        &err));
+    BOOST_CHECK(!VerifyScript(badsig5, scriptPubKey23, gFlags,
+                              TransactionSignatureChecker(
+                                  &txTo23, 0, txFrom23.vout[0].nValue, txdata),
+                              &err));
     BOOST_CHECK_MESSAGE(err == ScriptError::SIG_NULLFAIL,
                         ScriptErrorString(err));
 
     keys.clear(); // Must have signatures
     CScript badsig6 = sign_multisig(scriptPubKey23, keys, txTo23);
-    BOOST_CHECK(!VerifyScript(
-        badsig6, scriptPubKey23, gFlags,
-        TransactionSignatureChecker(&txTo23, 0, txFrom23.vout[0].nValue),
-        &err));
+    BOOST_CHECK(!VerifyScript(badsig6, scriptPubKey23, gFlags,
+                              TransactionSignatureChecker(
+                                  &txTo23, 0, txFrom23.vout[0].nValue, txdata),
+                              &err));
     BOOST_CHECK_MESSAGE(err == ScriptError::INVALID_STACK_OPERATION,
                         ScriptErrorString(err));
 }
@@ -2395,7 +2399,8 @@ BOOST_AUTO_TEST_CASE(script_combineSigs) {
     // Single signature case:
     BOOST_CHECK(SignSignature(keystore, CTransaction(txFrom), txTo, 0,
                               SigHashType().withForkId()));
-    scriptSig = DataFromTransaction(txTo, 0, txFrom.vout[0]);
+    scriptSig = DataFromTransaction(
+        txTo, 0, PrecomputedTransactionData(txTo, std::vector(txFrom.vout)));
     combined = CombineSignatures(txFrom.vout[0], txTo, scriptSig, empty);
     BOOST_CHECK(combined.scriptSig == scriptSig.scriptSig);
     combined = CombineSignatures(txFrom.vout[0], txTo, empty, scriptSig);
@@ -2404,7 +2409,8 @@ BOOST_AUTO_TEST_CASE(script_combineSigs) {
     // Signing again will give a different, valid signature:
     BOOST_CHECK(SignSignature(keystore, CTransaction(txFrom), txTo, 0,
                               SigHashType().withForkId()));
-    scriptSig = DataFromTransaction(txTo, 0, txFrom.vout[0]);
+    scriptSig = DataFromTransaction(
+        txTo, 0, PrecomputedTransactionData(txTo, std::vector(txFrom.vout)));
     combined =
         CombineSignatures(txFrom.vout[0], txTo, scriptSigCopy, scriptSig);
     BOOST_CHECK(combined.scriptSig == scriptSigCopy.scriptSig ||
@@ -2418,7 +2424,8 @@ BOOST_AUTO_TEST_CASE(script_combineSigs) {
     txTo.vin[0].prevout = COutPoint(txFrom.GetId(), 0);
     BOOST_CHECK(SignSignature(keystore, CTransaction(txFrom), txTo, 0,
                               SigHashType().withForkId()));
-    scriptSig = DataFromTransaction(txTo, 0, txFrom.vout[0]);
+    scriptSig = DataFromTransaction(
+        txTo, 0, PrecomputedTransactionData(txTo, std::vector(txFrom.vout)));
     combined = CombineSignatures(txFrom.vout[0], txTo, scriptSig, empty);
     BOOST_CHECK(combined.scriptSig == scriptSig.scriptSig);
     combined = CombineSignatures(txFrom.vout[0], txTo, empty, scriptSig);
@@ -2426,7 +2433,8 @@ BOOST_AUTO_TEST_CASE(script_combineSigs) {
     scriptSigCopy = scriptSig;
     BOOST_CHECK(SignSignature(keystore, CTransaction(txFrom), txTo, 0,
                               SigHashType().withForkId()));
-    scriptSig = DataFromTransaction(txTo, 0, txFrom.vout[0]);
+    scriptSig = DataFromTransaction(
+        txTo, 0, PrecomputedTransactionData(txTo, std::vector(txFrom.vout)));
     combined =
         CombineSignatures(txFrom.vout[0], txTo, scriptSigCopy, scriptSig);
     BOOST_CHECK(combined.scriptSig == scriptSigCopy.scriptSig ||
@@ -2438,7 +2446,8 @@ BOOST_AUTO_TEST_CASE(script_combineSigs) {
     BOOST_CHECK(keystore.AddCScript(scriptPubKey));
     BOOST_CHECK(SignSignature(keystore, CTransaction(txFrom), txTo, 0,
                               SigHashType().withForkId()));
-    scriptSig = DataFromTransaction(txTo, 0, txFrom.vout[0]);
+    scriptSig = DataFromTransaction(
+        txTo, 0, PrecomputedTransactionData(txTo, std::vector(txFrom.vout)));
     combined = CombineSignatures(txFrom.vout[0], txTo, scriptSig, empty);
     BOOST_CHECK(combined.scriptSig == scriptSig.scriptSig);
     combined = CombineSignatures(txFrom.vout[0], txTo, empty, scriptSig);
diff --git a/src/test/transaction_tests.cpp b/src/test/transaction_tests.cpp
index 73dad9c68..98dd1ce16 100644
--- a/src/test/transaction_tests.cpp
+++ b/src/test/transaction_tests.cpp
@@ -356,10 +356,11 @@ static void CheckWithFlag(const CTransactionRef &output,
                           bool success) {
     ScriptError error;
     CTransaction inputi(input);
+    const PrecomputedTransactionData txdata(inputi, {output->vout[0]});
     bool ret = VerifyScript(
         inputi.vin[0].scriptSig, output->vout[0].scriptPubKey,
         flags | SCRIPT_ENABLE_SIGHASH_FORKID,
-        TransactionSignatureChecker(&inputi, 0, output->vout[0].nValue),
+        TransactionSignatureChecker(&inputi, 0, output->vout[0].nValue, txdata),
         &error);
     BOOST_CHECK_EQUAL(ret, success);
 }
@@ -482,8 +483,10 @@ SignatureData CombineSignatures(const CMutableTransaction &input1,
                                 const CMutableTransaction &input2,
                                 const CTransactionRef tx) {
     SignatureData sigdata;
-    sigdata = DataFromTransaction(input1, 0, tx->vout[0]);
-    sigdata.MergeSignatureData(DataFromTransaction(input2, 0, tx->vout[0]));
+    sigdata = DataFromTransaction(
+        input1, 0, PrecomputedTransactionData(input1, {tx->vout[0]}));
+    sigdata.MergeSignatureData(DataFromTransaction(
+        input2, 0, PrecomputedTransactionData(input2, {tx->vout[0]})));
     const PrecomputedTransactionData txdata(*tx, {tx->vout[0]});
     ProduceSignature(DUMMY_SIGNING_PROVIDER,
                      MutableTransactionSignatureCreator(
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index bc9940bfa..78c0cce0f 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -2768,20 +2768,31 @@ bool CWallet::SignTransaction(CMutableTransaction &tx,
         }
     }
 
-    // At this point, one input was not fully signed otherwise we would have
-    // exited already Find that input and figure out what went wrong.
+    std::vector<CTxOut> spent_outputs(tx.vin.size());
+    std::vector<bool> has_coin(tx.vin.size());
     for (size_t i = 0; i < tx.vin.size(); i++) {
         // Get the prevout
         CTxIn &txin = tx.vin[i];
         auto coin = coins.find(txin.prevout);
         if (coin == coins.end() || coin->second.IsSpent()) {
             input_errors[i] = "Input not found or already spent";
+            // Just to be safe, sign unspendable if UTXO doesn't exist
+            spent_outputs[i].scriptPubKey = CScript() << OP_RETURN;
             continue;
         }
+        spent_outputs[i] = coin->second.GetTxOut();
+        has_coin[i] = true;
+    }
+    const PrecomputedTransactionData txdata(tx, std::move(spent_outputs));
 
+    // At this point, one input was not fully signed otherwise we would have
+    // exited already Find that input and figure out what went wrong.
+    for (size_t i = 0; i < tx.vin.size(); i++) {
+        if (!has_coin[i]) {
+            continue;
+        }
         // Check if this input is complete
-        SignatureData sigdata =
-            DataFromTransaction(tx, i, coin->second.GetTxOut());
+        SignatureData sigdata = DataFromTransaction(tx, i, txdata);
         if (!sigdata.complete) {
             input_errors[i] = "Unable to sign input, missing keys";
             continue;
-- 
2.39.5 (Apple Git-154)

